packages/wonder-stuff-core/src/kind-error.js

```diff
@@ -0,0 +1,125 @@
+// @flow
+import {Errors} from "./errors.js";
+import {ErrorInfo} from "./error-info.js";
+import {cloneMetadata} from "./clone-metadata.js";
+
+import type {Metadata} from "./types.js";
+
+type Options = {|
+    cause?: Error,
+    metadata?: $ReadOnly<Metadata>,
+    prefix?: string,
+    stripStackFrames?: number,
+    minimumFrameCount?: number,
+|};
+
+/**
+ * An error to describe detailed states and relationships.
+ *
+ * This error type supports error taxonomies and attachment of metadata, as
+ * well as causal relationships between itself and another source error.
+ * More specific variations can be built off this to provide a payload for
+ * logging setups like Sentry or winston.
+ */
+export class KindError extends Error {
+    +kind: string;
+    +originalMessage: string;
+    +metadata: ?$ReadOnly<Metadata>;
+
+    /**
+     * Creates an instance of `KindError`.
+     *
+     * @param {string} message The error message.
+     * @param {TKinds} [kind] The kind of error. This will be combined with
+     * `prefix` to form the name of the error, i.e. PrefixKindError.
+     * Defaults to `Errors.Unknown`.
+     * @param {Error} [cause] The error that caused this error.
+     * @param {Metadata} [metadata] The metadata to attach to the error.
+     * @param {string} [prefix] A prefix to prepend the name of the error.
+     * Defaults to `""`.
+     * @param {number} [stripStackFrames] The number of stack frames to remove
+     * from the error's stack. This can be used to ensure that the top call of
+     * the stack references the point at which an error is thrown which can
+     * be useful when helper functions are used to build the error being thrown.
+     * Defaults to `0`.
+     * @memberof KindError
+     */
+    constructor(
+        message: string,
+        kind: string = Errors.Unknown,
+        {
+            cause,
+            prefix,
+            metadata,
+            stripStackFrames,
+            minimumFrameCount,
+        }: Options = {
+            cause: undefined,
+            metadata: undefined,
+            prefix: "",
+            stripStackFrames: 0,
+            minimumFrameCount: 1,
+        },
```

Comment:
By using `Options`, it should make this much easier to build without needing helpers like `withSourceError` that the current `KAError` in mobile has added (which are great, yet feel like a workaround for the constructor lacking options like this).

Reply:
Also, you don't need the builder methods. You can always still use the constructor, but they're nice-to-haves, because I found that using the constructor, and thinking through the constructor args, was just taking too much effort, even though that extra effort was really tiny. Tiny was still too much, lol.

Reply:
That's why I used an options type rather than individual args. Provides the same level of discovery. There's nothing stopping derivations adding more things like helpers.

Reply:
I like the `Options` because then you don't have to worry about getting the order of the constructor arguments correct, and it supports multiple args being `null` and some not being `null`, without default values or other weirdnesses coming in to play. In that sense, it's totally solved half the problem!

It doesn't solve the problem of the developer having to discover the shape of what could be a very large object. The experience in my IDE might not be a nice, as the IDE attempts to suggest to me the arguments in-line as I type, because now they're down one level.

I'm not saying that we change _this_, just that we also still want builder methods!

Reply:
Oh wait, maybe context is too undefined... wonder what `Tags` looks like...

Reply:
Even if you type something like...

```
{
    "
}
```

Shame.

Reply:
But because `SentryTags` object is a defined flow-type, while in the auto-complete for `withTags`, I can click through to the definition/docs of `SentryTags`! I don't think it works on the docstrings for the individual properties though.
<img width="858" alt="Screen Shot 2021-10-14 at 2 44 44 PM" src="https://user-images.githubusercontent.com/185017/137400049-db39d19e-7fbb-4499-9617-54e2b90c55ef.png">